### ISSUE TITLE
bpf_probe_write_user helper function is locked down since linux kernel `5.14-rc6`

### ISSUE THREAD

**Issue by SzymonSt** [open-telemetry/opentelemetry-go-instrumentation#290]
### Describe the bug

Since this [commit](https://github.com/torvalds/linux/commit/51e1bb9eeaf7868db56e58f47848e364ab4c4129) in linux kernel repository, bpf_probe_write_user is locked down and this results in `unknown func bpf_probe_write_user` error during opentelemetry-go-instrumentation startup on systems with lockdown integrity enabled.

### Environment

 - OS: Ubuntu 22.04LTS Linux  - 5.15.0.1045.52-azure, (reproducible on all linux systems with kernel above 5.14 and integrity mode enabled)
 - Go Version: 1.20
 - Version: v0.2.1-alpha , v0.2.2-alpha

### To Reproduce

1. Build docker image from v0.2.1-alpha/v0.2.2-alpha
2. Run image on host with kernel version higher that 5.14

### Expected behavior

opentelemetry-go-instrumentation starts up without any issues.

### Additional context

bpf_probe_write_user should be probably retired from being used in opentelemetry-go-instrumentation as for many cloud providers/on demand use cases editing startup kernel parameters is not possible.

**Comment by SzymonSt** [comment]
Problem also  encountered and reported here: [237](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/237)

**Comment by damemi** [comment]
From 3-Mar-24 sig meeting, at @grcevski started a thread on the Linux kernel mailing list to see about unlocking this: https://www.uwsg.indiana.edu/hypermail/linux/kernel/2403.0/03026.html

It doesn't look like that thread has any replies yet, is there any way to follow up with that @grcevski ?

**Comment by damemi** [comment]
From SIG call today:

* We originally tried asking about this in Cillium slack, but didn't hear back
* Could possibly be fixed by LSM policies (are those usable in cloud providers though?)
* We should ask around eBPF/CNCF for advice or help with working on this in the kernel

**Comment by grcevski** [comment]
Based on this article, it seems that the lockdown LSM policies are static and cannot be modified or configured:

https://lwn.net/Articles/791863/

Proposals to make them more configurable appears to have been rejected.

**Comment by grcevski** [comment]
Relevant very new patch proposal https://lore.kernel.org/bpf/20240404190146.1898103-1-elver@google.com/

**Comment by grcevski** [comment]
Based on the latest comments on the thread I previously posted, it appears that the proposal for the new helper will not get accepted.

However, as of kernel 6.9 (to be released yet) there's a new feature called `bpf arena` (https://lwn.net/Articles/961594/). It will allow us to declare an arbitrary memory segment that can be shared between the userspace and the bpf program, allowing reads and writes. We'd have to confirm how this works yet, but if we were to make a launcher for a Go program, we can tap into the Go runtime and make any new GC allocated arena segment declared as `bpf arena`. If this works, we'd have full write access to the Go heap and eliminate the need for using the old helper. This would only work for newer kernels, but it would be a way forward.

**Comment by damemi** [comment]
Thanks @grcevski for the updates! I mentioned this on the call, but I think it's clear that we need to find an alternative approach. At this point I would consider the old helper DOA

**Comment by apconole** [comment]
I wonder if it would be possible instead to use a uprobe in some place like `go_crypto_tls_write` and `go_crypto_tls_read` to sniff the headers as they go along and use a bpf map that is shared to copy the headers around and create relationships to traces.  We may also be able to use these probes to at least monitor the related user headers (like any open telemetry key).  That's just on the uprobe side, because we want to trap data beyond the crypto boundary.  I'm not 100% on the exact probe points, but that's just from memory.

If we need to write into data structures, it may be possible to use a ptrace attached routine to push additional data in.  That will require permissions (as the current case) but I don't think it should be prevented by secureboot / lsm (but it is more restrictive because it is properly a debugging interface).  That's a bit more work, but I think it is doable.

**Comment by MrAlias** [comment]
@grcevski @RonFed ^

**Comment by grcevski** [comment]
> I wonder if it would be possible instead to use a uprobe in some place like `go_crypto_tls_write` and `go_crypto_tls_read` to sniff the headers as they go along and use a bpf map that is shared to copy the headers around and create relationships to traces. 

I apologize if I misunderstood the context, but this already happens, except the current code reads the headers information after the TLS decryption, in a TLS agnostic way. Essentially, he headers are read after they are parsed from the incoming request, regardless of TLS.

> If we need to write into data structures, it may be possible to use a ptrace attached routine to push additional data in. That will require permissions (as the current case) but I don't think it should be prevented by secureboot / lsm (but it is more restrictive because it is properly a debugging interface). That's a bit more work, but I think it is doable.

This is a very interesting idea, I don't think it will be limited by secureboot, since we already use ptrace to attach a shared memory segment to the instrumented process. I'd like to find out more about how you think this might work. If I understand correctly what you are saying, it will mean not using eBPF at all to inject the header values, but using a injected function hook to do the work?

**Comment by apconole** [comment]
> This is a very interesting idea, I don't think it will be limited by secureboot, since we already use ptrace to attach a shared memory segment to the instrumented process. I'd like to find out more about how you think this might work. If I understand correctly what you are saying, it will mean not using eBPF at all to inject the header values, but using a injected function hook to do the work?

Yes, exactly.

**Comment by grcevski** [comment]
After a lot of research I believe we might have a partial solution to this problem. There are two approaches we can implement here, which will give us partial coverage for the inability to write memory with `bpf_probe_write_user`:

### Approach without TLS
1. We'll need to start tracking the outgoing connection information when the Go code makes an outgoing server request. This can be achieved by setting up an uprobe on `net/http.(*persistConn).roundTrip`. At this point we'll capture the network tuple (source and destination ip address and source and destination port) and associate it to the current span context tracked by go auto instrumentation.
2. We'll add a new type of bpf program of sock_ops type, tracking only active established connections. Passive connections are for incoming requests and we don't need those. The sock_ops program will setup a sock hashmap with the kernel sock data structures. 
3. We'll add a new type of bpf program of sock_msg type, which will trigger on changes to socket operations tracked by the sock_ops map. Once a new packet is triggered, we will look up the information setup for us in step 1 and use `bpf_msg_push_data` to extend the packet for adding the `Traceparent` header.
4. We need one more program on Linux traffic control egress to write the actual value, using simple memcpy after calling `bpf_skb_pull_data`.

### Approach without TLS
1. This is a lot more hacky and it will require that we run go opentelemetry auto on both sides of the communication. But essentially we can use the Linux Traffic Control egress to store the context in the IP packet options, using the BPF helpers which allow us to extend the IP packets.
2. On Linux Traffic Control ingress we'll look at the IP options and extract the context and store it based on the network tuple.
3. The Go server uprobes will need to track the connection information and lookup the connection tuple context.

This is a lot more complex to what we do today, but it's a way forward. We discussed this approach with @damemi, @MrAlias  and some of the kernel maintainers at KubeCon NA 2024 and they suggested the approach described above without TLS is the right way to go.

**Comment by MrAlias** [comment]
> 1. This is a lot more hacky and it will require that we run go opentelemetry auto on both sides of the communication. But essentially we can use the Linux Traffic Control egress to store the context in the IP packet options, using the BPF helpers which allow us to extend the IP packets.

This will mean that the auto-instrumented service will not be a part of distributed traces with other non-auto-instrumented service for things like HTTPS.

Can we fallback to using bpf_probe_write_user when TLS is used?

Is KTLS still a possibility? It looks like there is a stalled proposal to add this to the Go stdlib: https://github.com/golang/go/issues/44506

**Comment by RonFed** [comment]
@grcevski Thanks for describing the solutions.

1.  Relating to @MrAlias comment, assuming `bpf_probe_write_user` is available, I'd prefer to keep the current approach and avoid loading extra programs (regardless if TLS is used), WDYT?
2. Are there possible scenarios where multiple active spans will use the same network tuple? (gRPC streams are one example that I can think of) - in that case, we would need to add another field to the map key, right?
3. Currently we don't read kernel structures, so I think we don't require the presence of BTF, will this become a requirement? I don't know how rare it is for BTF to be disabled or not, and how is it compared to a secure boot being on or off.

**Comment by JamieDanielson** [comment]
👋  Looks like this may have stalled. I am hearing reports of more folks having trouble using the auto-instrumentation in the operator because of lockdown integrity enabled.

>one way would be to add an #ifdef around the calls to bpf_probe_write_user and build different versions with the bpf2go at compile time. Then all we need to do is add logic in the userspace to detect kernel >= 5.14 + integrity mode and load the safe version of the bpf probes.

This check was suggested higher in the thread and seems like it might be a reasonable starting point to unblock folks now, and allow time to determine the best long term solution for this scenario. What do you think?

**Comment by JamieDanielson** [comment]
> load the safe version of the bpf probes.

To clarify, I think the safe version if possible might be disabling context propagation entirely for folks with lockdown integrity, if it means allowing startup and instrumentation.

**Comment by grcevski** [comment]
Yes you are right, I will work on a PR. The work has been done for HTTP, but we need to apply the same approach for gRPC and Kafka.

**Comment by apconole** [comment]
NOTE: There is a new effort underway upstream (https://lore.kernel.org/all/20250305-afabre-traits-010-rfc2-v1-0-d0ecfb869797@cloudflare.com/) and this may be a way to add tracing data that can be propagated across various security boundaries.  There is some effort to get a proper uapi integration for this so that userspace tools (reverse proxies) can preserve this data as well.  It might be the right approach for this.

**Comment by damemi** [comment]
@apconole thanks for the update! Using the extra header space sounds like something @grcevski had suggested before, which is interesting

**Comment by RonFed** [comment]
https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/2105 - avoiding to stop the instrumentation in case context propagation is not supported for `kafka-go`.

**Comment by MrAlias** [comment]
SIG meeting note:

- Still need to check gRPC

**Comment by diegocejasprieto** [comment]
Hello, is there any update on this? I'm having this issue relate to auto-instrumentation while running a Go app on EKS

`{"time":"2025-04-22T18:46:28.366027703Z","level":"ERROR","source":{"function":"go.opentelemetry.io/auto/internal/pkg/instrumentation.(*Manager).loadProbes","file":"/app/internal/pkg/instrumentation/manager.go","line":377},"msg":"error while loading probes, cleaning up","error":"program uprobe_writeSubset: load program: invalid argument: unknown func bpf_probe_write_user#36 (910 line(s) omitted)","name":{"SpanKind":3,"InstrumentedPkg":"net/http"}}
{"time":"2025-04-22T18:46:28.366088289Z","level":"DEBUG","source":{"function":"go.opentelemetry.io/auto/internal/pkg/instrumentation/probe.(*Base[...]).Close","file":"/app/internal/pkg/instrumentation/probe/probe.go","line":256},"msg":"Closed","Probe":{"SpanKind":3,"InstrumentedPkg":"net/http"}}
{"time":"2025-04-22T18:46:28.942348254Z","level":"DEBUG","source":{"function":"go.opentelemetry.io/auto/internal/pkg/instrumentation/probe.(*Base[...]).Close","file":"/app/internal/pkg/instrumentation/probe/probe.go","line":256},"msg":"Closed","Probe":{"SpanKind":2,"InstrumentedPkg":"net/http"}}
{"time":"2025-04-22T18:46:28.942485027Z","level":"DEBUG","source":{"function":"go.opentelemetry.io/auto/internal/pkg/instrumentation.(*Manager).cleanup","file":"/app/internal/pkg/instrumentation/manager.go","line":409},"msg":"Cleaning bpffs"}
{"time":"2025-04-22T18:46:28.94263693Z","level":"ERROR","source":{"function":"main.main","file":"/app/cli/main.go","line":125},"msg":"failed to load instrumentation","error":"program uprobe_writeSubset: load program: invalid argument: unknown func bpf_probe_write_user#36 (910 line(s) omitted)"`

**Comment by prgsing** [comment]
Hi, I see this issue with AKS nodes

`{"time":"2025-08-11T19:19:37.109854067Z","level":"INFO","source":{"function":"go.opentelemetry.io/auto/internal/pkg/instrumentation.(*Manager).loadProbes","file":"/usr/src/go.opentelemetry.io/auto/internal/pkg/instrumentation/manager.go","line":382},"msg":"loading probe","name":{"SpanKind":3,"InstrumentedPkg":"google.golang.org/grpc"}}
{"time":"2025-08-11T19:19:37.254753648Z","level":"ERROR","source":{"function":"go.opentelemetry.io/auto/internal/pkg/instrumentation.(*Manager).loadProbes","file":"/usr/src/go.opentelemetry.io/auto/internal/pkg/instrumentation/manager.go","line":385},"msg":"error while loading probes, cleaning up","error":"program uprobe_LoopyWriter_HeaderHandler: load program: invalid argument: unknown func bpf_probe_write_user#36 (245 line(s) omitted)","name":{"SpanKind":3,"InstrumentedPkg":"google.golang.org/grpc"}}
{"time":"2025-08-11T19:19:38.381753406Z","level":"ERROR","source":{"function":"main.main","file":"/usr/src/go.opentelemetry.io/auto/cli/main.go","line":175},"msg":"failed to load instrumentation","error":"program uprobe_LoopyWriter_HeaderHandler: load program: invalid argument: unknown func bpf_probe_write_user#36 (245 line(s) omitted)"}

Any workaround to solve this?
`

**Comment by woehrl01** [comment]
For completeness, on newer kernels the error seems to be a bit different: 

e.g. running on latest bottlerocket os

```
program uprobe_writeSubset: load program: invalid argument: program of this type cannot use helper bpf_probe_write_user#36 (726 line(s) omitted)
```
